// @medusajs/medusa/dist/models/customer-group.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomerGroup = void 0;
var typeorm_1 = require("typeorm");
var ulid_1 = require("ulid");
var __1 = require("..");
var db_aware_column_1 = require("../utils/db-aware-column");
var CustomerGroup = /** @class */ (function () {
    function CustomerGroup() {
    }
    CustomerGroup.prototype.beforeInsert = function () {
        if (this.id) {
            return;
        }
        var id = (0, ulid_1.ulid)();
        this.id = "cgrp_".concat(id);
    };
    __decorate([
        (0, typeorm_1.PrimaryColumn)(),
        __metadata("design:type", String)
    ], CustomerGroup.prototype, "id", void 0);
    __decorate([
        (0, typeorm_1.Index)({ unique: true, where: "deleted_at IS NULL" }),
        (0, typeorm_1.Column)(),
        __metadata("design:type", String)
    ], CustomerGroup.prototype, "name", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return __1.Customer; }, function (customer) { return customer.groups; }, {
            onDelete: "CASCADE",
        }),
        (0, typeorm_1.JoinTable)({
            name: "customer_group_customers",
            joinColumn: {
                name: "customer_group_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "customer_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], CustomerGroup.prototype, "customers", void 0);
    __decorate([
        (0, typeorm_1.CreateDateColumn)({ type: (0, db_aware_column_1.resolveDbType)("timestamptz") }),
        __metadata("design:type", Date)
    ], CustomerGroup.prototype, "created_at", void 0);
    __decorate([
        (0, typeorm_1.UpdateDateColumn)({ type: (0, db_aware_column_1.resolveDbType)("timestamptz") }),
        __metadata("design:type", Date)
    ], CustomerGroup.prototype, "updated_at", void 0);
    __decorate([
        (0, typeorm_1.DeleteDateColumn)({ type: (0, db_aware_column_1.resolveDbType)("timestamptz") }),
        __metadata("design:type", Date)
    ], CustomerGroup.prototype, "deleted_at", void 0);
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({ type: "jsonb", nullable: true }),
        __metadata("design:type", Object)
    ], CustomerGroup.prototype, "metadata", void 0);
    __decorate([
        (0, typeorm_1.BeforeInsert)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], CustomerGroup.prototype, "beforeInsert", null);
    CustomerGroup = __decorate([
        (0, typeorm_1.Entity)()
    ], CustomerGroup);
    return CustomerGroup;
}());
exports.CustomerGroup = CustomerGroup;
