// @medusajs/medusa-js/dist/resources/admin/draft-orders.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var qs_1 = __importDefault(require("qs"));
var base_1 = __importDefault(require("../base"));
var AdminDraftOrdersResource = /** @class */ (function (_super) {
    __extends(AdminDraftOrdersResource, _super);
    function AdminDraftOrdersResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @description Creates a draft order
     */
    AdminDraftOrdersResource.prototype.create = function (payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    /**
     * @description Add line item to draft order
     */
    AdminDraftOrdersResource.prototype.addLineItem = function (id, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id + "/line-items";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    /**
     * @description Delete draft order
     */
    AdminDraftOrdersResource.prototype.delete = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id;
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    /**
     * @description Remove line item
     */
    AdminDraftOrdersResource.prototype.removeLineItem = function (id, itemId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id + "/line-items/" + itemId;
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    /**
     * @description Retrieves a draft order
     */
    AdminDraftOrdersResource.prototype.retrieve = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id;
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    /**
     * @description Lists draft orders
     */
    AdminDraftOrdersResource.prototype.list = function (query, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders";
        if (query) {
            var queryString = qs_1.default.stringify(query);
            path = "/admin/draft-orders?" + queryString;
        }
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    /**
     * @description Mark a draft order as paid
     */
    AdminDraftOrdersResource.prototype.markPaid = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id + "/pay";
        return this.client.request("POST", path, {});
    };
    /**
     * @description Update draft order
     */
    AdminDraftOrdersResource.prototype.update = function (id, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id;
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    /**
     * @description Update draft order line item
     */
    AdminDraftOrdersResource.prototype.updateLineItem = function (id, itemId, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/draft-orders/" + id + "/line-items/" + itemId;
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    return AdminDraftOrdersResource;
}(base_1.default));
exports.default = AdminDraftOrdersResource;
