// @medusajs/medusa-js/dist/resources/admin/invites.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = __importDefault(require("../base"));
var AdminInvitesResource = /** @class */ (function (_super) {
    __extends(AdminInvitesResource, _super);
    function AdminInvitesResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminInvitesResource.prototype.accept = function (payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/invites/accept";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminInvitesResource.prototype.create = function (payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/invites";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminInvitesResource.prototype.delete = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/invites/" + id;
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    AdminInvitesResource.prototype.list = function (customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/invites";
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    AdminInvitesResource.prototype.resend = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/invites/" + id;
        return this.client.request("POST", path, {}, {}, customHeaders);
    };
    return AdminInvitesResource;
}(base_1.default));
exports.default = AdminInvitesResource;
