// gatsby-source-medusa/operations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOperations = void 0;
const client_1 = require("./client");
function createOperations(options, { reporter }) {
    const client = (0, client_1.createClient)(options, reporter);
    function createOperation(name, queryString) {
        return {
            execute: () => client[name](queryString),
            name: name,
        };
    }
    return {
        createProductsOperation: createOperation("products"),
        createCollectionsOperation: createOperation("collections"),
        createRegionsOperation: createOperation("regions"),
        createOrdersOperation: createOperation("orders"),
        incrementalProductsOperation: (date) => createOperation("products", date.toISOString()),
        incrementalCollectionsOperation: (date) => createOperation("collections", date.toISOString()),
        incrementalRegionsOperation: (date) => createOperation("regions", date.toISOString()),
        incrementalOrdersOperation: (date) => createOperation("orders", date.toISOString()),
    };
}
exports.createOperations = createOperations;
